### PR TITLE
Gitlab support

### PR DESCRIPTION
**Pull request by Swahjak**
First off; Awesome job on this satis implementation. Have been struggling with Satis for way to long and this solves everything that is wrong with the 'default' satis implementation.

I have been following along since I first encountered this repository and created a branch to support Gitlab. Thought you might be interested in checking it out and possibly merging it.

### CODE REVIEW

**Review by francoispluchino — CHANGES_REQUESTED** [review]
Thank you for your message and your PR. I could not answer you before, because I was in vacation in a place where I did not have any network connection ... It is always a pleasure to see my work being appreciated and used. Add support to Gitlab is something I wanted to do, so thanks for your help (just like the support for native GIT). However, I have some requests before merging your PR:

- Rebase your PR into 1 commit
- Fix your tests

File: aws/cloud-formation-stack.yaml [hunk 1]
```diff
@@ -1,7 +1,6 @@
 AWSTemplateFormatVersion: '2010-09-09'
 Transform: AWS::Serverless-2016-10-31
 Description: Serverless Composer repository for private PHP packages
-
```
Comment (francoispluchino) [inline aws/cloud-formation-stack.yaml]: Remove this empty line

File: src/composer/repositories/VcsRepository.ts [hunk 2]
```diff
@@ -92,8 +94,6 @@ export class VcsRepository
         }
 
         this.driver = new this.drivers[validType](this.repoData, this.config) as VcsDriver;
-        this.repoData['type'] = 'vcs-' + validType;
```
Comment (francoispluchino) [inline src/composer/repositories/VcsRepository.ts]: Do not delete these 2 lines.

File: src/composer/repositories/vcs-drivers/GitlabDriver.ts [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+import {Config} from '../../../configs/Config';
+import {VcsDriverError} from '../../../errors/VcsDriverError';
+import {RemoteFilesystem} from '../../utils/RemoteFilesystem';
+import {VcsDriver} from './VcsDriver';
+import {TransportError} from "../../../errors/TransportError";
+import querystring from 'querystring';
+import {LooseObject} from '../../../utils/LooseObject';
+
+/**
+ * @author François Pluchino <francois.pluchino@gmail.com>
```
Comment (francoispluchino) [inline src/composer/repositories/vcs-drivers/GitlabDriver.ts]: Replace by your full name and email, it's your work :-).

File: src/configs/Config.ts [hunk 4]
```diff
@@ -16,9 +16,12 @@ import merge from 'lodash.merge';
 export class Config
 {
     public static defaultConfig: LooseObject = {
-        'github-domains': ['github.com'],
```
Comment (francoispluchino) [inline src/configs/Config.ts]: Do not delete this line.

**Review by francoispluchino — CHANGES_REQUESTED** [review]
Thanks to make changes of your code for comments on each line :-).

**Comment by Swahjak** [comment]
@francoispluchino No problem. I will try to look into it in the upcoming week.

**Comment by Swahjak** [comment]
@francoispluchino I'm sorry haven't got to it before my holidays. I will try to make time next week.

**Comment by francoispluchino** [comment]
@Swahjak Thank you for your contribution, however, your PR was still awaiting modifications, and in the meantime, a lot of changes have taken place to arrive at the first stable version. Therefore, your PR is in conflict, and I so added personally this new functionality via the commit d85390b34fc058aa820ee7c6cbb058bf061cdb32.

**Comment by Swahjak** [comment]
@francoispluchino I can only applaud that :clap:

**Pull request closed without merging**